lightning_pose/data/utils.py

```diff
@@ -284,15 +284,40 @@ def generate_heatmaps(
     return confidence
 
 
+# @typechecked
+# def evaluate_heatmaps_at_location(
+#     heatmaps: TensorType["batch", "num_keypoints", "heatmap_height", "heatmap_width"],
+#     locs: TensorType["batch", "num_keypoints", 2],
+# ) -> TensorType["batch", "num_keypoints"]:
+#     """Evaluate 4D heatmaps using a 3D location tensor (last dim is x, y coords)."""
+#     i = torch.arange(heatmaps.shape[0]).reshape(-1, 1, 1, 1)
+#     j = torch.arange(heatmaps.shape[1]).reshape(1, -1, 1, 1)
+#     k = locs[:, :, None, 1, None].type(torch.int64)  # y first
+#     l = locs[:, :, 0, None, None].type(torch.int64)  # x second
+#     vals = heatmaps[i, j, k, l].squeeze(-1).squeeze(-1)  # get rid of singleton dims
+#     return vals
+
 @typechecked
 def evaluate_heatmaps_at_location(
     heatmaps: TensorType["batch", "num_keypoints", "heatmap_height", "heatmap_width"],
     locs: TensorType["batch", "num_keypoints", 2],
+    sigma: Union[float, int] = 1.25,  # sigma used for generating heatmaps
 ) -> TensorType["batch", "num_keypoints"]:
     """Evaluate 4D heatmaps using a 3D location tensor (last dim is x, y coords)."""
     i = torch.arange(heatmaps.shape[0]).reshape(-1, 1, 1, 1)
     j = torch.arange(heatmaps.shape[1]).reshape(1, -1, 1, 1)
-    k = locs[:, :, None, 1, None].type(torch.int64)  # y first
-    l = locs[:, :, 0, None, None].type(torch.int64)  # x second
-    vals = heatmaps[i, j, k, l].squeeze(-1).squeeze(-1)  # get rid of singleton dims
-    return vals
+    k = locs[:, :, None, 1, None].type(torch.int64)
+    l = locs[:, :, 0, None, None].type(torch.int64)
+    pix_to_consider = int(np.ceil(sigma * 2.0)) #get all pixels within two stds.
+    offsets = list(np.arange(-pix_to_consider, pix_to_consider+1))
+    vals_all = []
+    for offset in offsets:
```

Comment:
looks good, height and width may differ (i.e., we may have a non square heatmap). so the max clamping should be `heatmap_height-1` then `heatmap_width-1`. you currently use `max=heatmaps.shape[2]-1` for both. other than that, fine

Reply:
Fixed, thanks!